gcc/rust/expand/rust-macro-builtins.cc

```diff
@@ -252,4 +263,42 @@ MacroBuiltin::compile_error (Location invoc_locus, AST::MacroInvocData &invoc)
   return AST::ASTFragment::create_error ();
 }
 
+/* Expand builtin macro concat!(), which joins all the literal parameters
+   into a string with no delimiter. */
+
+AST::ASTFragment
+MacroBuiltin::concat (Location invoc_locus, AST::MacroInvocData &invoc)
+{
+  auto invoc_token_tree = invoc.get_delim_tok_tree ();
+  MacroInvocLexer lex (invoc_token_tree.to_token_stream ());
+  Parser<MacroInvocLexer> parser (std::move (lex));
+  auto str = std::string ();
+
+  auto last_token_id = macro_end_token (invoc_token_tree, parser);
+
+  /* NOTE: concat! could accept no argument, so we don't have any checks here */
+  while (parser.peek_current_token ()->get_id () != last_token_id)
+    {
+      auto token = parser.peek_current_token ();
+      if (token->is_literal () || token->get_id () == TRUE_LITERAL
+	  || token->get_id () == FALSE_LITERAL)
+	{
+	  auto lit_expr = parser.parse_literal_expr ();
+	  str += lit_expr->as_string ();
+	}
+      else
+	{
+	  rust_error_at (parser.peek_current_token ()->get_locus (),
+			 "argument must be a constant literal");
```

Comment:
Rustc also accepts a macro invocation here which should expand to a string.

Reply:
I think this is related to #1085. All the built-in macros in Rust GCC currently do not support nesting macro expressions.